### PR TITLE
DAOS-3324 test: Adding Fio test infrastructure and a CI test.

### PR DESCRIPTION
**Pull request by saurabhtandan**
As part of this PR FIO test infrastructure is being added
using fio_utils.py, fio_test_base.py.
Also adding ftest/io/fio_small.py and ftest/io/fio_small.yaml
for CI testing.

Change-Id: I5b212c64f1dc7219ffea3dbdcb528f274c53d040
Signed-off-by: Saurabh Tandan <saurabh.tandan@intel.com>

### CODE REVIEW

File: src/tests/ftest/io/fio_small.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            1/8 Clients
+            num of files/dirs: 100
+            with/without unique working dir for each task
+            write bytes: 0|4K
+            read bytes: 0|4K
+            depth of hierarchical directory structure: 0|5
+        :avocado: tags=all,pr,hw,fio,fiosmall
+        """
+#        fio_flags = self.params.get("flags", "/run/fio/*")
+#        self.fio_cmd.flags.update(fio_flags)
```
Comment (mjean308) [inline src/tests/ftest/io/fio_small.py]: Is this needed?

File: src/tests/ftest/util/fio_utils.py [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+
+           Args:
+             host: hostname from where to launch fio
+        """
+        # run fio command
+        self.run_cmd = self.run_cmd.replace("fio", ' ').replace(" POSIX", '')
+        ret_code = general_utils.pcmd(host, 'fio' + self.run_cmd)
+
+        # check for any failures
+        if 0 not in ret_code:
```
Comment (Puspalogan) [inline src/tests/ftest/util/fio_utils.py]: It is best to convert the ret_code to integer and check "if not ret_code". There are non-zero values that have "0" in them which will skip this block while it shouldn't.

Comment (phender) [reply src/tests/ftest/util/fio_utils.py]: The `ret_code` object is a dictionary of integer keys with `NodeSet` values (the list of hosts that yielded the return code that matches the key).  So `<int> not in <dict>` where the keys are of type `int` is the efficient way to do this.

Comment (Puspalogan) [reply src/tests/ftest/util/fio_utils.py]: Agreed. Initially I thought ret_code has single string value. Later, Saurabh explained me that it was a dict and it makes sense to keep the code this way.

File: src/tests/ftest/io/fio_small.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        Use Cases:
+            Aim of this test is to test different combinations
+            of following configs:
+            1/8 Clients
+            num of files/dirs: 100
+            with/without unique working dir for each task
+            write bytes: 0|4K
+            read bytes: 0|4K
+            depth of hierarchical directory structure: 0|5
+        :avocado: tags=all,pr,hw,fio,fiosmall
```
Comment (brianjmurrell) [inline src/tests/ftest/io/fio_small.py]: Does this **need** to be run on hardware?

Comment (saurabhtandan) [reply src/tests/ftest/io/fio_small.py]: Yes it does.

Comment (brianjmurrell) [reply src/tests/ftest/io/fio_small.py]: Why is that?

Comment (saurabhtandan) [reply src/tests/ftest/io/fio_small.py]: So fio is another benchmark like ior but we can perform many different io patterns using FIO on our underlying nvme and dcpm storage. Hence this one needs to be on hardware. The question was if we should keep it on pr or full_regression. But Johann confirmed, we should have it on 'pr' run.

File: src/tests/ftest/io/fio_small.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+    def test_fio_small(self):
+        """Jira ID: DAOS-2493.
+        Test Description:
+            Test Fio in small config.
+        Use Cases:
+            Aim of this test is to test different combinations
+            of following configs:
+            1/8 Clients
+            num of files/dirs: 100
```
Comment (mchaarawi) [inline src/tests/ftest/io/fio_small.py]: did you just copy this text from mdtest? that doesn't look right to me..

File: src/tests/ftest/util/fio_utils.py [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+
+class FioCommand(ExecutableCommand):
+    """Defines a object representing a fio command."""
+
+    def __init__(self, namespace, command, test):
+        """Create a fio Command object."""
+        super(FioCommand, self).__init__(namespace, command)
+        self.test = test
+
+        # fio commandline options
```
Comment (mchaarawi) [inline src/tests/ftest/util/fio_utils.py]: why do you need to add all this here?

File: src/tests/ftest/io/fio_small.yaml [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        scm_size: 1600000000
+        nvme_size: 9000000000
+    createsvc:
+        svcn: 1
+fio:
+  global:
+    name: 'global'
+    directory: "/tmp/daos_dfuse"
+    ioengine: 'libaio'
+    thread: 16
```
Comment (mchaarawi) [inline src/tests/ftest/io/fio_small.yaml]: what does that mean (setting it to 16)?
AFAIK it's 1 or 0; use threads or use processes. unless that changed in a recent FIO release..

File: src/tests/ftest/io/fio_small.yaml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    group_reporting: 1
+    direct: 1
+    verify: 'crc64'
+    iodepth: 32
```
Comment (mchaarawi) [inline src/tests/ftest/io/fio_small.yaml]: i would suggest to put this at 16

**Pull request merged**